feat: Implement full-text search

PSkinnerTech opened the pull request:
Enhanced the search functionality to index and search full text content within the docs pages. 

This includes:

- Indexing all text content, including paragraphs, code blocks, and lists.
- Searching across both titles and content with appropriate weighting.
- Improving search result relevance by considering context and prioritizing title matches.

Bobinstein commented:
@diempi Thank you for pointing this out. 
We ended up correcting the issue by making sure the url resolved without the `.html`, so these updates are not necessary.

diempi commented:
You are welcome Bob ☺️

PSkinnerTech committed:
feat: Implement full-text search

Enhanced the search functionality to index and search full text content within documentation pages. This includes:

- Indexing all text content, including paragraphs, code blocks, and lists.
- Searching across both titles and content with appropriate weighting.
- Improving search result relevance by considering context and prioritizing title matches.

This update provides more comprehensive search results, allowing users to find information more effectively.

Bobinstein commented:
This seems to successfully search the full text of each page, however, the displayed search results are not very informative.
I had envisioned something closer to what is implemented in the vuepress docs portal, which shows both the page name and section name where the search result is found, and if it is a result from the full text, the surrounding text is also displayed in the search result so users have more context when selecting a result.

https://github.com/ar-io/docs/blob/main/docs/src/.vuepress/theme/components/SearchModal.vue

PSkinnerTech committed:
feat: enhance search with full-text content and improved result display

- Implement full-text search across all document content
- Add preview text to search results for better context
- Display content type (text/code) in search results
- Show complete hierarchy path (section/page/subsection)
- Improve search result relevance with proper content weighting

This enhances the search functionality to match the capabilities
of the v1 docs portal while providing better context in results.

PSkinnerTech commented:
# Enhanced Search with Full-Text Content and Improved Result Display

## Changes
- Implemented full-text search across all document content, not just headers
- Added preview text to search results showing the matching content in context
- Improved the result display to show complete hierarchy (section/page/subsection)
- Added content type indicators to differentiate between text and code blocks
- Improved search result relevance by properly weighting different content types

## Technical Details
- Modified the MDX processor to capture and index all content types
- Enhanced the search index configuration to store additional context
- Added proper type definitions for improved result format

## Screenshots
<img width="1782" alt="Screenshot 2024-12-17 at 16 41 15" src="https://github.com/user-attachments/assets/d369bc25-03b8-45da-b93b-a8dcec0cc008" />
